### PR TITLE
[BrowserKit] Add support for HttpClient

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | part of #30502
| License       | MIT
| Doc PR        | not yet

When combining the power of the new HttpClient component with the BrowserKit and Mime components, we can makes something really powerful... a full/better/awesome replacement for https://github.com/FriendsOfPHP/Goutte.

So, this PR is about integrating the HttpClient component with BrowserKit to give users a high-level interface to ease usages in the most common use cases.

Scraping websites can be done like this:

```php
use Symfony\Component\BrowserKit\HttpBrowser;
use Symfony\Component\HttpClient\HttpClient;

$client = HttpClient::create();
$browser = new HttpBrowser($client);

$browser->request('GET', 'https://example.com/');
$browser->clickLink('Log In');
$browser->submitForm('Sign In', ['username' => 'me', 'password' => 'pass']);
$browser->clickLink('Subscriptions')->filter('table tr:nth-child(2) td:nth-child(2)')->each(function ($node) {
    echo trim($node->text())."\n";
});
```

And voilà! Nice, isn't?

Want to add HTTP cache? Sure:

```php
use Symfony\Component\HttpKernel\HttpCache\Store;

$client = HttpClient::create();
$store = new Store(sys_get_temp_dir().'/http-cache-store');

$browser = new HttpBrowser($client, $store);

// ...
```

Want logging and debugging of HTTP Cache? Yep:

```php
use Psr\Log\AbstractLogger;

class EchoLogger extends AbstractLogger
{
    public function log($level, $message, array $context = [])
    {
        echo $message."\n";
    }
}

$browser = new HttpBrowser($client, $store, new EchoLogger());
```

The first time you run your code, you will get an output similar to:

```
Request: GET https://twig.symfony.com/
Response: 200 https://twig.symfony.com/
Cache: GET /: miss, store
Request: GET https://twig.symfony.com/doc/2.x/
Response: 200 https://twig.symfony.com/doc/2.x/
Cache: GET /doc/2.x/: miss, store
```

But then:

```
Cache: GET /: fresh
Cache: GET /doc/2.x/: fresh
```

Limit is the sky here as you get the full power of all the Symfony ecosystem.

Under the hood, these examples leverage HttpFoundation, HttpKernel (with HttpCache),
DomCrawler, BrowserKit, CssSelector, HttpClient, Mime, ...

Excited?

P.S. : Tests need to wait for the HttpClient Mock class to land into master.

### CODE REVIEW

File: src/Symfony/Component/BrowserKit/README.md [hunk 1]
```diff
@@ -4,6 +4,9 @@ BrowserKit Component
 The BrowserKit component simulates the behavior of a web browser, allowing you
 to make requests, click on links and submit forms programmatically.
 
+The component comes with a concreate implementation that uses the HttpClient
```
Comment (nlubisch) [inline src/Symfony/Component/BrowserKit/README.md]: I think "concreate" should be "concrete"

**Review by nlubisch — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/BrowserKit/HttpBrowser.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+use Symfony\Component\Mime\Part\TextPart;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * An implementation of a browser using the HttpClient component
+ * to make real HTTP requests.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class HttpBrowser extends AbstractBrowser
```
Comment (ro0NL) [inline src/Symfony/Component/BrowserKit/HttpBrowser.php]: `WebBrowser` maybe?

**Review by ro0NL — COMMENTED** [review]
> Limit is the sky

You mean "Sky's the limit" i suppose :D

File: src/Symfony/Component/BrowserKit/HttpBrowser.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/BrowserKit/HttpBrowser.php]: This consumes an `HttpClientInterface`, so `Http` makes sense to me - `Web` doesn't look as specific to me. Dunno what others think.

Comment (ro0NL) [reply src/Symfony/Component/BrowserKit/HttpBrowser.php]: I have no real pref :) just wondering if `HTTP Browser` is a "real world" thing ... (https://en.wikipedia.org/wiki/Web_browser vs. https://en.wikipedia.org/wiki/HTTP_browser)

Comment (fabpot) [reply src/Symfony/Component/BrowserKit/HttpBrowser.php]: I tend to agree with @ro0NL here, especially as this class is optimized for browsing websites (and not APIs).

File: src/Symfony/Component/BrowserKit/HttpBrowser.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        $this->store = $store;
+        $this->logger = $logger ?? new NullLogger();
+
+        parent::__construct();
+    }
+
+    protected function doRequest($request)
+    {
+        if ($this->store) {
+            return $this->doCachedHttpRequest($request);
```
Comment (thewilkybarkid) [inline src/Symfony/Component/BrowserKit/HttpBrowser.php]: Are there plans to have a `CachingHttpClient`?

File: src/Symfony/Component/BrowserKit/HttpBrowser.php (hunk 2 above)
Comment (stof) [reply src/Symfony/Component/BrowserKit/HttpBrowser.php]: @ro0NL HTTP qualifies this implementation of the Browser, as opposed to KernelBrowser and HttpKernelBrowser.

*all* browser implementations of BrowserKit are about web browsers (but we haven't added `web` everywhere in the component)

File: src/Symfony/Component/BrowserKit/HttpBrowser.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/BrowserKit/HttpBrowser.php]: plans no, but you know how Symfony works: contributions fuel it forward ;)

**Review by nicolas-grekas — APPROVED** [review]
Green in a minute, with tests provided by @ktherage. The caching part is now separated into #30629, and logging has been removed. It should be reintroduced in `HttpClient` directly.

File: src/Symfony/Component/BrowserKit/HttpBrowser.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * to make real HTTP requests.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @final
+ */
+class HttpBrowser extends AbstractBrowser
+{
+    private $client;
+    private $httpKernelBrowser;
```
Comment (ro0NL) [inline src/Symfony/Component/BrowserKit/HttpBrowser.php]: unused

Comment (nicolas-grekas) [reply src/Symfony/Component/BrowserKit/HttpBrowser.php]: removed thanks

**Review by chalasr — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you THERAGE Kévin.

**Pull request merged**